nginx.conf

```diff
@@ -99,6 +99,16 @@ http {
 
       location ~ /id_/follow_redirect/(?<proxied_uri>.*) {
         set $upstream $proxied_uri$is_args$args;
+        # Return 405 Not Allowed when attempting to proxy blacklisted urls.
+        # - 10.0.0.0/8      RFC1918 possible internal network
+        # - 172.16.0.0/12   RFC1918 possible internal network
+        # - 192.168.0.0/16  RFC1918 possible internal network
+        # - 169.254.0.0/16  RFC3927 link-local addresses (including AWS instance metadata URL)
+        # - 127.0.0.0/8     RFC3330 internal loopback block
+        if ( $proxied_uri ~ ^.*//(10\.(([0-9])|([1-9][0-9])|(1[0-9][0-9])|(2(([0-4][0-9])|(5[0-5]))))\.(([0-9])|([1-9][0-9])|(1[0-9][0-9])|(2(([0-4][0-9])|(5[0-5]))i))\.(([0-9])|([1-9][0-9])|(1[0-9][0-9])|(2(([0-4][0-9])|(5[0-5])))))|(172\.((1[6-9])|([2-9][0-9])|(1[0-9][0-9])|(2([0-4][0-9])|(5[0-5])))\.(([0-9])|([1-9][0-9])|(1[0-9][0-9])|(2(([0-4][0-9])|(5[0-5]))))\.(([0-9])|([1-9][0-9])|(1[0-9][0-9])|(2(([0-4][0-9])|(5[0-5])))))|(192\.168\.(([0-9])|([1-9][0-9])|(1[0-9][0-9])|(2(([0-4][0-9])|(5[0-5]))))\.(([0-9])|([1-9][0-9])|(1[0-9][0-9])|(2(([0-4][0-9])|(5[0-5])))))|(169\.254\.(([0-9])|([1-9][0-9])|(1[0-9][0-9])|(2(([0-4][0-9])|(5[0-5]))))\.(([0-9])|([1-9][0-9])|(1[0-9][0-9])|(2(([0-4][0-9])|(5[0-5])))))|(127\.(([0-9])|([1-9][0-9])|(1[0-9][0-9])|(2(([0-4][0-9])|(5[0-5]))))\.(([0-9])|([1-9][0-9])|(1[0-9][0-9])|(2(([0-4][0-9])|(5[0-5]))))\.(([0-9])|([1-9][0-9])|(1[0-9][0-9])|(2(([0-4][0-9])|(5[0-5]))))).*$ ) {
```

Comment:
Hi Hannah,

I'm afraid this approach looks quite problematic to me:

- It is very hard to verify that the lengthy regex here matches the comment above, and adding a new address is going to be a nightmare
- The regex is duplicated below. It would be hard to tell if they got out of sync.
- The `$proxied_uri` argument here, if I understand correctly, is untrusted user input, not a resolved IP address? This means that a malicious actor can bypass the check by finding an alternative way of constructing a URL which resolves to the forbidden IPs. From a quick Google, there appear to be at least 4 alternative ways of representing any given IP: https://www.silisoftware.com/tools/ipconverter.php?convert_from=127.0.0.1. I haven't even thought about whether IPv6 is an issue here. The existing squid-based setup is [not fooled by alternative representations of IPv4 addresses](https://via.hypothes.is/http://2852039166/latest/dynamic/).

Reply:
If there's a better way to do this I haven't found one. This is generally why I like the approach of modifying the host files to block access vs sorting through a string in nginx. It's possible there's some module out there but I haven't found it. If we use the host config the only ip's we need to block here are the ones running inside the docker container which makes this list much shorter and much more manageable.

Reply:
See [slack discussion](https://hypothes-is.slack.com/archives/C1MA4E9B9/p1572278534017000). Going to go back to using iptables and keep this list scoped to docker internal ips.